### DISCUSSION TITLE
How it support actions for vue in storybook?

### DISCUSSION THREAD

**Discussion by wanhuihan** [storybookjs/storybook#11372] [General]
I tried several times for this issue, found that mostly it support react.js, not vue.js.
I want to make the click event work in actions, how to solve it? thanks.

**Comment by shilman** [comment]
AFAIK actions works for all frameworks including vue. you can see it working in our vue example apps e.g. https://next--storybookjs.netlify.app/vue-kitchen-sink/

**Comment by lostdesign** [comment]
- Install and add `'@storybook/addon-actions',` to your `main.js` inside the `addons: []`
- Import the action in your `foo.stories.js` -> `import { action } from '@storybook/addon-actions'`
- In the template, have a `@click="someMethod"` on your component
- Add a `methods: {}` object like so: `methods: { someMethod: action('clicked') }`

**Example**
```js
import { action } from '@storybook/addon-actions'

import HelloWorld from './hello-world'

export default {
  title: 'Atoms / Hello World',
}

export const Default = () => ({
  components: { HelloWorld },
  template:
    '<hello-world @click="action"></hello-world>',
  methods: { action: action('clicked') }
})
```

**Reply by ghost** [reply]
Note to self: Need to add this to docs https://storybook.js.org/docs/vue/essentials/actions

**Reply by putskan** [reply]
Is this considered best practice? 
@lostdesign 
@3nuc

**Reply by lostdesign** [reply]
I would say it is still pretty much valid to do like so. When I wrote my original answer, SB was still in beta.

**Reply by putskan** [reply]
Thanks for the quick reply!
Is there a good way to use that approach to handle mutiple event types, without duplicate code?

For example, handling @ click, @ mouseover together:

```
import { action } from "@storybook/addon-actions"
import InfoCard from "../components/InfoCard.vue"

export default {
  title: "InfoCard",
  component: InfoCard,
}

const Template = args => ({
  components: { InfoCard },
  methods: { action },
  data: () => ({ args }),
  template: `
    <info-card 
    @click="action('click')()" 
    @mouseover="action('mouseover')()"
    v-bind="args" />
    `,
})

**Reply by lostdesign** [reply]
I don't quite understand the issue?

**Reply by anhance** [reply]
It works, but number of actions didn't show up, unfortunately. Maybe that's an issue with vue 2?

**Reply by benjamin-richardson-circular** [reply]
@anhance one thing that tripped me up was that the `action()` is a factory, where it just returns another function that you need to call. I am also using Vue 2 for this.

To make this a little more neat, I defined my action functions at the top, then call them in the code.

```js
export default {
  title: "InfoCard",
  component: InfoCard,
}

const clickAction = action('click');
const mouseoverAction = actions('mouseover');

const Template = args => ({
  components: { InfoCard },
  methods: { action },
  data: () => ({ args }),
  template: `
    <info-card 
    @click="clickAction()" 
    @mouseover="mouseoverAction()"
    v-bind="args" />
    `,
})
```

**Reply by yandongxu** [reply]
> * Install and add `'@storybook/addon-actions',` to your `main.js` inside the `addons: []`
> * Import the action in your `foo.stories.js` -> `import { action } from '@storybook/addon-actions'`
> * In the template, have a `@click="someMethod"` on your component
> * Add a `methods: {}` object like so: `methods: { someMethod: action('clicked') }`
> 
> **Example**
> 
> ```js
> import { action } from '@storybook/addon-actions'
> 
> import HelloWorld from './hello-world'
> 
> export default {
>   title: 'Atoms / Hello World',
> }
> 
> export const Default = () => ({
>   components: { HelloWorld },
>   template:
>     '<hello-world @click="action"></hello-world>',
>   methods: { action: action('clicked') }
> })
> ```

try this:

```javascript
import HelloWorld from './hello-world'

export default {
  title: 'Atoms / Hello World',
  // Add `argTypes` config
  argTypes: {
    click: {
      action: 'click',
    },
  },
}

export const Default = (_, { argTypes }) => ({
  components: { HelloWorld },
  props: Object.keys(argTypes),
  // Magic!
  template: '<hello-world v-bind="$props" @click="click"></hello-world>',
})
```

**Comment by tokidoki11** [comment]
i tried the things same as written here

https://storybook.js.org/blog/improved-type-safety-in-storybook-7/


<img width="458" alt="image" src="https://github.com/storybookjs/storybook/assets/7564844/b0f4358b-1154-4cd5-96ef-e19b41c9e8a8">

<img width="386" alt="スクリーンショット 2024-03-01 10 12 19" src="https://github.com/storybookjs/storybook/assets/7564844/872a6dd2-12ba-4e5f-a865-9a9906e6252d">

but event is not showing in args or argtypes

**Reply by eljefedelrodeodeljefe** [reply]
Thanks for that hint. Worked for me with Some adaptations also with SB 8. I guess moving the storybook to TS was also helpful.

A note though: based on [this](https://storybook.js.org/blog/storybook-test/) the pattern probably should be using `fn` from `@storybook/test`

```typescript
import { fn } from '@storybook/test'
import type { Meta } from '@storybook/vue3'

export default {
  title: 'Example/ChatContainer',
  component: ChatContainer,
  tags: ['autodocs'],
  args: {
    currentThreadId: '1',
    onOnMessagesLoadMoreRequested: fn(() => {}),
    onOnSearchInput: fn(() => {}),
    onOnNewThreadRequested: fn(() => {})
  }
} satisfies Meta<typeof ChatContainer>
```
